src/vtlengine/Operators/Time.py

```diff
@@ -810,14 +825,92 @@ def validate(cls, Date1: Any, Date2: Any) -> Any:
+            if error is not None:
+                raise SemanticError(f"2-1-19-{error}",
+                                    op=cls.op,
+                                    type=param.__class__.__name__ if error == 12 else
+                                    param.data_type.__name__,
+                                    name="shiftNumber" if error == 12 else "periodInd",
+                                    expected="Scalar" if error == 12 else expected_types[i].__name__
+                                    )
+
+        if (isinstance(operand, (Scalar, DataComponent)) and
+                operand.data_type not in [Date, TimePeriod]):
+            unary_implicit_promotion(operand.data_type, Date)
+
+        if isinstance(operand, Scalar):
+            return Scalar(name=operand.name, data_type=operand.data_type, value=None)
+        if isinstance(operand, DataComponent):
+            return DataComponent(name=operand.name, data_type=operand.data_type, data=None)
+
+        if all(comp.data_type not in [Date, TimePeriod] for comp in operand.components.values()):
+            raise SemanticError("2-1-19-14", op=cls.op, name=operand.name)
+        return Dataset(name='result', components=operand.components.copy(), data=None)
+
+    @classmethod
+    def evaluate(cls,
+                 operand: Union[Scalar, DataComponent, Dataset],
+                 param_list: List[Scalar]
+                 ) -> Union[Scalar, DataComponent, Dataset]:
+        result = cls.validate(operand, param_list)
+        shift, period = param_list[0].value, param_list[1].value
+        is_tp = isinstance(operand, (Scalar, DataComponent)) and operand.data_type == TimePeriod
+
+        if isinstance(result, Scalar) and isinstance(operand, Scalar) and operand.value is not None:
+            result.value = cls.op_date(operand.value, shift, period, is_tp)
+        elif (isinstance(result, DataComponent) and isinstance(operand, DataComponent) and
+              operand.data is not None):
+            result.data = operand.data.map(lambda x: cls.op_date(x, shift, period, is_tp),
+                                           na_action="ignore")
+        elif (isinstance(result, Dataset) and isinstance(operand, Dataset) and
+              operand.data is not None):
+            result.data = operand.data.copy()
+            for measure in operand.get_measures():
+                if measure.data_type in [Date, TimePeriod]:
+                    result.data[measure.name] = result.data[measure.name].map(
+                        lambda x: cls.op_date(x, shift, period, measure.data_type == TimePeriod),
+                        na_action="ignore")
+                    measure.data_type = Date
+
+        if isinstance(result, (Scalar, DataComponent)):
+            result.data_type = Date
+        return result
 
     @classmethod
-    def validate(cls, operand: Any, param_list: List[Any]) -> Any:
-        pass
+    def op_date(cls,
```

Comment:
Maybe change the name to py_op for consistency with other operators?

Reply:
Changed!